inmoose/utils/splines.py

```diff
@@ -0,0 +1,100 @@
+
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+# -----------------------------------------------------------------------------
+
+# original code was contributed to StackOverflow: https://stackoverflow.com/questions/71550468/does-python-have-an-analogue-to-rs-splinesns
+# documentation was added
+
+# As per StackOverflow terms of service (see https://stackoverflow.com/help/licensing), original code is licensed under CC BY SA 4.0, which is compatible with GPL3 (see https://creativecommons.org/2015/10/08/cc-by-sa-4-0-now-one-way-compatible-with-gplv3/)
+
+
+import numpy as np
+from scipy.interpolate import splev
+
+
+def ns(x, df=None, knots=None, boundary_knots=None, include_intercept=False):
+    """
+    Generate the B-spline basis matrix for a natural cubic spline.
+
+    This function intends to provide the same functionality as R splines::ns.
+
+    Arguments
+    ---------
+    x : array_like
+        the predictor variable
+    df : int, optional
+        degrees of freedom. If :code:`knots` is not specified, then the
+        function chooses :code:`df - 1 - intercept` knots at suitably chosen
+        quantiles of :code:`x`. If :code:`None`, the number of inner knots is
+        set to :code:`len(knots)`.
+    knots : array_like, optional
+        breakpoints that define the spline. The default is no knots; together
+        with the natural boundary conditions this results in a basis for linear
+        regression on :code:`x`. Typical values are the mean or median for one
+        knot, quantiles for more knots. See also :code:~boundary_knots`.
+    include_intercept : bool, optional
+        if :code:`True`, an intercept is included in the basis; default is
+        :code:`False`.
+    boundary_knots : array_like, optional
+        boundary points at which to impose the natural boundary conditions and
+        anchor the B-spline basis (default the range of the data). If both
+        :code:`knots` and :code:`boundary_knots` are supplied, the basis
+        parameters do not depend on :code:`x`. Data can extend beyond
+        :code:`boundary_knots`.
+
+    Returns
+    -------
+    ndarray
+        a matrix of dimension :code:`(len(x), df)`, where :code:`df =
+        len(knots)-1-intercept` if :code:`df` was not supplied.
+    """
+    degree = 3
+    if boundary_knots is None:
+        boundary_knots = [np.min(x), np.max(x)]
+
+    if df is not None:
+        nIknots = df - 1 - include_intercept
+        if nIknots < 0:
+            nIknots = 0
+
+        if nIknots > 0:
+            knots = np.linspace(0, 1, num=nIknots + 2)[1:-1]
+            knots = np.quantile(x, knots)
```

Comment:
What happends if knots is None and `nIknots > 0`?
It looks like knots is not redefined?

Reply:
`knots` may remain `None`, in which case the spline is determined based on boundary conditions only.